R/data.table.R

```diff
@@ -2815,3 +2764,136 @@ gforce <- function(env, jsub, o, f, l, rows) .Call(Cgforce, env, jsub, o, f, l,
 isReallyReal <- function(x) {
   .Call(CisReallyReal, x)
 }
+
+.prepareFastSubset <- function(isub, x, enclos, notjoin, verbose = FALSE){
+  ## helper that decides, whether a fast binary search can be performed, if i is a call
+  ## Fast binary search is currently possible for calls like colA == 1, colA %in% c(1,2)
+  ## colA == 1 & colB == 2 & colC %chin% c("a", "b"), ...
+  ## Notjoin is only supported for single column subsets.
+  ## Additional restrictions are imposed if x is .SD, or if options indicate that no optimization 
+  ## is to be performed
+  #' @param isub the substituted i
+  #' @param x the data.table
+  #' @param enclos The environment where to evaluate when RHS is not a column of x
+  #' @param notjoin boolean that is set before, indicating whether i started with '!'
+  #'                If notjoin == TRUE, only calls without '&' connection are supported, i.e. single column subsets
+  #' @param verbose TRUE for detailed output
+  #' @return If i is not fast subsettable, NULL. Else, a list with entries: 
+  #'        out$i: a data.table that will be used as i with proper column names and key.
+  #'        out$xo: the ordering index that will be used by bmerge (integer(0) for no reordering necesary)
+  #'        out$leftcols The column numbers in i
+  #'        out$rightcols The corresponding column numbers in x
+  #'        out$notjoin Bool. In some cases, notjoin is updated within the function.
+  #'        ATTENTION: If nothing else helps, an auto-index is created on x unless options prevent this.
+  if (!is.call(isub)) return(NULL)
+  if (!is.null(attr(x, '.data.table.locked'))) return(NULL)  # fix for #958, don't create auto index on '.SD'.
+  if (!getOption("datatable.use.index")) {
+    return(NULL) # #1422 
+    ## Does this option also prevent the usage of keys? It is interpreted in this way here...
+  }
+  ## Determine, whether the nature of isub in general supports fast binary search
+  remainingIsub <- isub
+  i <- list()
```

Comment:
I would use more meaningful name for this variable `i` might be easily confused with `[.data.table` `i` arg.

Reply:
actually, this is the list that will become the `i` arg for `bmerge` after it is `CJ`ed. Therefore, I think the naming makes sense.